mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/DeleteMessageListener.java

```diff
@@ -0,0 +1,145 @@
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+public class DeleteMessageListener implements MailboxListener.GroupMailboxListener {
+    private static final Optional<CassandraId> ALL_MAILBOXES = Optional.empty();
+
+    public static class DeleteMessageListenerGroup extends Group {
+
+    }
+
+    @Inject
+    public DeleteMessageListener(CassandraMessageIdToImapUidDAO imapUidDAO, CassandraMessageIdDAO messageIdDAO, CassandraMessageDAO messageDAO,
+                                 CassandraAttachmentDAOV2 attachmentDAO, CassandraAttachmentOwnerDAO ownerDAO,
+                                 CassandraAttachmentMessageIdDAO attachmentMessageIdDAO) {
+        this.imapUidDAO = imapUidDAO;
+        this.messageIdDAO = messageIdDAO;
+        this.messageDAO = messageDAO;
+        this.attachmentDAO = attachmentDAO;
+        this.ownerDAO = ownerDAO;
+        this.attachmentMessageIdDAO = attachmentMessageIdDAO;
+    }
+
+    private final CassandraMessageIdToImapUidDAO imapUidDAO;
+    private final CassandraMessageIdDAO messageIdDAO;
+    private final CassandraMessageDAO messageDAO;
+    private final CassandraAttachmentDAOV2 attachmentDAO;
+    private final CassandraAttachmentOwnerDAO ownerDAO;
+    private final CassandraAttachmentMessageIdDAO attachmentMessageIdDAO;
+
+    @Override
+    public Group getDefaultGroup() {
+        return new DeleteMessageListenerGroup();
+    }
+
+    @Override
+    public boolean isHandling(Event event) {
+        return event instanceof Expunged || event instanceof MailboxDeletion;
+    }
+
+    @Override
+    public void event(Event event) {
+        if (event instanceof Expunged) {
+            Expunged expunged = (Expunged) event;
+
+            Flux.fromIterable(expunged.getExpunged()
+                .values())
+                .map(MessageMetaData::getMessageId)
+                .map(CassandraMessageId.class::cast)
+                .concatMap(this::handleDeletion)
+                .then()
+                .block();
+        }
+        if (event instanceof MailboxDeletion) {
+            MailboxDeletion mailboxDeletion = (MailboxDeletion) event;
+
+            CassandraId mailboxId = (CassandraId) mailboxDeletion.getMailboxId();
+
+            messageIdDAO.retrieveMessages(mailboxId, MessageRange.all())
+                .map(ComposedMessageIdWithMetaData::getComposedMessageId)
+                .concatMap(metadata -> imapUidDAO.delete((CassandraMessageId) metadata.getMessageId(), mailboxId)
+                    .then(messageIdDAO.delete(mailboxId, metadata.getUid()))
+                    .then(handleDeletion((CassandraMessageId) metadata.getMessageId())))
+                .then()
+                .block();
+        }
+    }
+
+    private Mono<Void> handleDeletion(CassandraMessageId messageId) {
+        return Mono.just(messageId)
+            .filterWhen(this::isReferenced)
+            .flatMap(id -> readMessage(id)
```

Comment:
can we avoid that reading by putting more data in the event?

Reply:
Event deseriazation logic will become tricky. 

Can come in a second step IMO (as an optimization)

Reply:
> Event deseriazation logic will become tricky.

Why?

Reply:
Events JSON is persistend in DeadLetter and the eventBus.

Either you have to require them to be empty upon upgrade (what had been refused last time I try to change event format) or you need to support retro-compatibility in the event format.

Reply:
Anyway, I will let your team deal with it ;-)

Reply:
> Events JSON is persistend in DeadLetter and the eventBus.

Adding fields should not be a problem. BTW, we can't live forever stuck with the current format, we'll need to handle this kind of cases

> Anyway, I will let your team deal with it ;-)

I don't understand. 
And which team are you talking about?

Reply:
> I don't understand.
> And which team are you talking about?

The Linagora Moka transformer team obviously.

I was not aware it was a ticket on your board.

> Events JSON is persistend in DeadLetter and the eventBus.

Modifying the event format to save reads here is clearly an optimization that can come in a follow up work.

Reply:
> Modifying the event format to save reads here is clearly an optimization that can come in a follow up work.

Yes.

Just a last question: is MessageRepresentation immutable?

Reply:
> Just a last question: is MessageRepresentation immutable?

Yes it is immutable